---
title: "Analysis of BLAST Results"
author: "Maggie Chen"
date: "October 12, 2018"
output: github_document
---

# Introduction
  
This is a further analysis of the research performed by researchers from University of South Columbia, who intend to investigate whether microbe community on human skins will match the microbe community on objects touched by the same individual. Microbe community refers to the microorganisms composition in a shared living space. In this research, the main focus is on the bacterial community composition. Bacteria is a type of biological cell, and it is a member of unicellular microorganisms group. Bacteria have various shape and length, as well as different effect on human bodies. The research was also intend to use this method as an additional tool of forensic examination. Forensic investigation is denoting the application of scientific methods and techniques to the investigation of crime. Thus, it is possible to match the microbe community on suspect's hand with the object the suspect has touched as one form of forensic evidence. Overall, the sample swabbing, extraction, amplification, and sequencing was done by researchers from University of South Columbia at University of Colorado. Then, using the sequence data and the metadata, I was able to run several more analysis on the sequence and use BLAST to identify species from the sequence. The metadata is the information about the sequence including the information about sampling individuals and sampling conditions. Basic Local Alignment Sequence Tool(BLAST), is a method using alignment to match portions in the sequence of interest with the sequence in the database, which is already unknown for species names. 

# Methods
## Sample origin and sequencing
  
For sampling, in the same building at the University of Colorado, five healthy male and four healthy female in between the age of 20 and 35 were recruited for sampling. The entire surface of each computer mouse and the palm surface of the individual's dominant hand, which is commonly used to operate the mouse was swabbed. To ensure the sampling quality the mouse has been touched by the owner within 12h before the swabbing. Palm swabbing was done while each individual remained their typical hand hygiene practices. All swab samples were stored at -80 ÂºC before DNA extraction.  
  
The DNA extraction was performed using the MO BIO PowerSoil DNA isolation kit. For each sample, the primers were optimized for the phylogenetic analysis of pyrosequencing reads. Then PCR reaction was used to amplify the sequence. If the final sequence length is less than 200 or more than 300bp in length, or had a quality score lower than 25, or contained either ambiguous character or correctable barcode, or lacked primer sequence, it was removed from the analysis. 

## Computational

In addition to the analysis performed in the research project, I also ran several analysis on the sequence data. Using the data gained from the sequencing in the research. Firstly, I ran a quality check on all the sequences using fastqc, which generates a QC report with quality evaluations of the sequence based on different parameters like GC content, the regional quality score of the sequence. Then, based on the quality score of the sequence, I trimmed out the low quality (lower than 25) portions in each sequence. So far, I have only working on the sequence in the format of '.fasta'. However, the final goal for this sequence processing is to use BLAST to identify different species within each sequence, which only accepts '.fastq' file. Thus, I used 'biowak' to convert all the trimmed sequences into '.fastq' format. Lastly, I input all the converted sequences into BLAST, to identify all the species in the sample taking by researches. Overall, to process this analysis faster, the work was done on a remote machine with better capability than personal laptops. To analyze the BLAST result, as a primary interpretation, codes are written to count and summarize the species names found. For further analysis, R is used to generate different graphs, which will be demonstrated in the later portion of this file.

# Results
In order to further analyze the BLAST data, graphs were generated to illustrate different patterns and interesting findings in the data. Firstly, 'pident' is a abbreviation of percent identity, which presents what percentage the sequence being found matches with the sequence in the existing sequence in GenBank. The lower the percentage, the less reliable the match is. For the bacteria species found on male palm and male touched mouse, there is a difference in sequence matching qualities. 'dust' refers to the sample taken from the mouse, 'sebum' refers to the data taken from human palm (Figure 2). Sequences from male touched mouse has a good matching quality, in which most of the sample sequence has 100% match. The samples of male palm has a variation of percent identity, some are at 100% and 93%. However, most abundant are at 85%. Most sequences from female palm sample has a high confident to match with the existing sequences in GenBank, which is the opposite with male samples (Figure 2 and 3). 
  
Almost all the samples from female palm have 100% identity. On the other hand, samples from female touched mouse sample has concentrated at 85% identity, with some at 93% and 100% identity. I counted the repeating numbers of each species that has percent identity between 85% and 87% (Figure 1). For species that have percent identity between 85% and 87%, *Solemya pervernicosa gill symbiont* is the species that has the highest times being identified in BLAST(Figure 1). In both groups of mouse and palm samples, by the separation of female and male, *Solemya pervernicosa gill symbiont* is most frequently found species overall, and only found in male palm and female mouse (Figure 4). *Bartonella washoensis* is the second commonly found species that is only found in male mouse and female palm. *Acidovorax sp.* is only found on both female and male palm. *Pinus oocarpa* and *Staphylococcus succinus* are only found on the male mouse. In table1. it is a summary of the amount of species found in each individual's samples, including the repeating species.

###Graphs
```{r load-libraries, message = FALSE, echo=FALSE}
# Be sure to install these packages before running this script
# They can be installed either with the install.packages() function
# or with the 'Packages' pane in RStudio

# load packages
library("dplyr")
library("tidyr")
library("knitr")
library("ggplot2")
```

```{r make-read-in-data-function, echo = FALSE}
# Output format from BLAST is as detailed on:
# https://www.ncbi.nlm.nih.gov/books/NBK279675/
# In this case, we used: '10 sscinames std'
# 10 means csv format
# sscinames means unique Subject Scientific Name(s), separated by a ';'
# std means the standard set of result columns, which are:
# 'qseqid sseqid pident length mismatch
# gapopen qstart qend sstart send evalue bitscore',


# this function takes as input a quoted path to a BLAST result file
# and produces as output a dataframe with proper column headers
# and the 'qseqid' column split into sample and seq number
read_blast_output <- function(filename) {
  data_in <- read.csv(filename,
                      header = FALSE, # files don't have column names in them
                      col.names = c("sscinames", # unique Subject Sci Name(s)
                                    "qseqid",    # Query Seq-id
                                    "sseqid",    # Subject Seq-id
                                    "pident",    # Percntge of identical matches
                                    "length",    # Alignment length
                                    "mismatch",  # Number of mismatches
                                    "gapopen",   # Number of gap openings
                                    "qstart",    # Start of alignment in query
                                    "qend",      # End of alignment in query
                                    "sstart",    # Start of alignment in subj
                                    "send",      # End of alignment in subject
                                    "evalue",    # Expect value
                                    "bitscore"))  # Bit score

  # Next we want to split the query sequence ID into
  # Sample and Number components so we can group by sample
  # They originally look like "ERR1942280.1"
  # and we want to split that into two columns: "ERR1942280" and "1"
  # we can use the separate() function from the tidyr library to do this
  # Note that we have to double escape the period for this to work
  # the syntax is
  # separate(column_to_separate,
  # c("New_column_name_1", "New_column_name_2"),
  # "seperator")
  data_in <- data_in %>%
    separate(qseqid, c("sample_name", "sample_number"), "\\.")
}
```

```{r read-in-BLAST-data, echo = FALSE}
# this makes a vector of all the BLAST output file names, including
# the name(s) of the directories they are in
files_to_read_in <- list.files(path = "output/blast",
                               full.names = TRUE)

# We need to create an empty matrix with the right number of columns
# so that we can rbind() each dataset on to it
joined_blast_data <- matrix(nrow = 0,
                            ncol = 14)

# now we loop over each of the files in the list and append them
# to the bottom of the 'joined_blast_data' object
# we do this with the rbind() function and the function we
# made earlier to read in the files, read_blast_output()
for (filename in files_to_read_in) {
  joined_blast_data <- rbind(joined_blast_data,
                             read_blast_output(filename))
}
```

```{r read-in-metadata-and-join, echo = FALSE}
# Next we want to read in the metadata file so we can add that in too
# This is not a csv file, so we have to use a slightly different syntax
# here the "sep = "\t"" tells the function that the data are tab-delimited
# and the "stringsAsFactors = FALSE" tells it not to assume that things are
# categorical variables
metadata_in <- read.table(paste0("data/metadata/",
                                 "fierer_forensic_hand_mouse_SraRunTable.txt"),
                          sep = "\t",
                          header = TRUE,
                          stringsAsFactors = FALSE)

# Finally we use the left_join() function from dplyr to merge or 'join' the
# combined data and metadata into one big table, so it's easier to work with
# in R the "by = c("Run_s" = "sample_name")" syntax tells R which columns
# to match up when joining the datasets together
joined_blast_data_metadata <- metadata_in %>%
  left_join(joined_blast_data,
            by = c("Run_s" = "sample_name"))
```


```{r group-by-summarize, echo = FALSE}
# group by anonyamized name and calculate mean percent identity
# filter the mean pident to the range of 85%-87%
# pipe the results to a new data set
# count the times of each species repeated
# filter to only show species repeated more than 3 times
joined_blast_data_metadata %>%
  group_by(sscinames) %>%
  summarize(mean_pident = mean(pident),
            sd_pident = sd(pident)) %>%
  filter(mean_pident > 85) %>%
  filter(mean_pident < 87) %>%
  select(sscinames) %>%
  distinct() %>%
  pull() -> target_pident_spp
  
joined_blast_data_metadata %>%
  filter(sscinames %in% target_pident_spp) %>%
  group_by(sscinames) %>%
  tally() %>%
  arrange(desc(n)) %>%
  filter(n > 3)
```

Figure 1. This figure to illustrate percent identity in the range between 85% and 87%, about how many times each species are identifies. The most abundant match is *Solemya pervernicosa gill symbiont*, then *Geosporobacter sp.* IRF9 with a significant lower repetition times. This figure is only showing the species that has more than 3 times of repetition.

```{r histogram-male, echo = FALSE}
# using the dplyr piping syntax to select male using 'M'
# in the row of  host_subject_id_s, and then pull out the column
# of both dust and sebum from the data to make a histogram.
joined_blast_data_metadata %>%
   filter(grepl("M", host_subject_id_s)) %>%
  ggplot(aes(x = pident)) +
    geom_histogram() +
    ggtitle("Male Percent Identity") +
    xlab("Percent") +
  facet_grid(env_material_s ~ .)
```

Figure 2. Sequences from male touched mouse has a good matching quality, in which most of the sample sequence has 100% match. The samples of male palm, has a variation of percent identity, some are at 100% and 93%. However, the most abundant are at 85%.  


```{r histogram-female, echo = FALSE}
# using the dplyr piping syntax to select female using 'F'
# in the row of  host_subject_id_s, and then pull out the column of
# dust and sebum from the data to make a histogram
# used facet_grid to catagorize samples into
# dust and human skin(sebum), and make 2 seperate graphs
joined_blast_data_metadata %>%
   filter(grepl("F", host_subject_id_s)) %>%
  ggplot(aes(x = pident)) +
    geom_histogram() +
    ggtitle("Female Percent Identity") +
    xlab("Percent") +
    facet_grid(env_material_s ~ .)
```

Figure 3. Almost all the samples from female plum has 100% identity. Samples from female touched mouse sample has concentrated at 85% identity, with some at 93% and 100% identity.


```{r female-male-speceis-counting, echo = FALSE}
# Count number of each species found and the corrolation with host
# filter the species in each sex, dust or sebum that repeated
# more than 155 times to narrow out the top 8
# most commonly found species
joined_blast_data_metadata %>%
  mutate(sex_all_samples = substring(anonymized_name_s, 1, 1)) %>%
  group_by(sscinames, sex_all_samples, env_material_s) %>%
  tally() %>%
  arrange(desc(n)) %>%
  filter(n > 155) %>%
  ggplot(aes(x = sscinames,
             y = n,
             fill = sex_all_samples)) +
  geom_col(position = position_dodge()) +
  theme(axis.text.x = element_text(angle = 90,
                                   hjust = 1)) +
   ggtitle("Most frquently found species
           count in each sex and each locations") +
  xlab("species name") +
  ylab("number found") +
  facet_grid(env_material_s ~ .)
```

Figure 4. In both groups of mouse and plum samples, by the separation of female and male, the top 8 most frequently found species. *Solemya pervernicosa gill symbiont* is most frequently found species overall, and only found in male palm and female mouse. *Bartonella washoensis* is the second commonly found species that is only found in male mouse and female palm. *Acidovorax sp.* is only found on both female and male palm. *Pinus oocarpa* and *Straphylococcus succinus* are only found on male mouse. 


```{r summary-table, echo = FALSE}
# Finally, we'd like to be able to make a summary table of the counts of
# sequences for each subject for both sample types. To do that we can use the
# table() function. We add the kable() function as well (from the tidyr package)
# in order to format the table nicely when the document is knitted
kable(table(joined_blast_data_metadata$host_subject_id_s,
            joined_blast_data_metadata$sample_type_s))
```

Table1. A summary of amount of species found in each individual's samples, including the repeating species. 

# Discussion
The percent identity variation was observed in both female and male sample, however, in the opposite patterns. The high percent identify for most of the samples was in the female palm and male mouse, which is not observed in female mouse and male palm samples. One possible explanation is that the species Solemya pervernicosa gill symbiont as showed in Figure1 is the primary cause of this variation. According to Figure3, Solemya pervernicosa gill symbiont was only found in female mouse and male palm sample. Also, due to more than two thousand Solemya pervernicosa gill symbiont was identified in all the sample and it has relatively low percent identity (around 85%). Thus, the peak around 85% percent identity was generated only with female mouses and male palm samples, while female palm and male mouse was not affected. Solemya pervernicosa gill symbiont was not found in both male palm and male mouse, same for females. This inconsistency requires more experiments for further investigation. In Table1 the amount of species identifies in each individual's palm and mouse does not match, this might due to some species has a high number of repetition, which is not necessarily generated by identification of different species. However, this difference also indicates there are some other factors affect the microbe community composition. 
   
Overall, all the samples and identifications are in acceptable quality, indicated by the percent identity except the *Solemya pervernicosa gill symbiont*. There are several possible reasons why the *Solemya pervernicosa gill symbiont* has a high identification rate with low percent identity. One is that the species found in the sample are not the same species within the BLAST database. In this case, a more detail analysis is needed to determine whether the specie found is Solemya pervernicosa gill symbiont or not. Another possible reason is that there are parts of the sequence got trimmed out due to the low sequence quality, due to the in-completion and some mismatch, it then has a low percent identify. One thing to note is that mismatch does not always yield dysfunction or mutation of the organism. 
   
As for the biological significance of the species found in samples. *Solemya pervernicosa gill symbiont*'s common name is gammaproteobacteria. Proteobacteria is a major phylum of gram-negative bacteria, which include a wide range of pathogens. Bartonella washoeensis belongs to the genus of bartonella, which is first isolate from dog with mistral valve endocarditis, then it was also found on the human with cardiac disease(Bruno ,etc. 2018). This can also be an indication of pets' disease or bacteria that are related to the human disease can be passed on from daily lives. *Straphylococcus succinus* is a gram-positive bacteria, which is found on cheese, sausages and healthy wild animal skins, it was also isolated from clinical material. However, its parthenogenesis has yet to investigate. *Pinus oocarpa* is a species of pine tree native to Mexico and Central America. In summary, the biological significance of each species being identified was not clear yet. There is a possibility that the sample sequence is too short thus it matches an irrelevant species in the database. 
 
In conclusion, there is very less correlation between the microbe community on human hands and object touched by the same individual. The most commonly found species has a variant location and function. Also, the biological significance of each species found needs further investigation.  





